driver/src/orderbook/mod.rs

```diff
@@ -27,26 +29,47 @@ pub trait StableXOrderBookReading {
     fn get_auction_data(&self, index: U256) -> Result<(AccountState, Vec<Order>)>;
 }
 
-/// A simple implementation of the order book trait that forwards calls
-/// directly to the smart contract.
-pub struct StableXOrderBookReader<'a> {
+/// Implements the StableXOrderBookReading trait by using the underlying
+/// contract in a paginated way.
+/// This avoid hitting gas limits when the total amount of orders is large.
+pub struct PaginatedStableXOrderBookReader<'a, 'b> {
     contract: &'a dyn StableXContract,
+    page_size: u64,
+    web3: &'b Web3,
 }
 
-impl<'a> StableXOrderBookReader<'a> {
-    pub fn new(contract: &'a dyn StableXContract) -> Self {
-        Self { contract }
+impl<'a, 'b> PaginatedStableXOrderBookReader<'a, 'b> {
+    pub fn new(contract: &'a dyn StableXContract, page_size: u64, web3: &'b Web3) -> Self {
+        Self {
+            contract,
+            page_size,
+            web3,
+        }
     }
 }
 
-impl<'a> StableXOrderBookReading for StableXOrderBookReader<'a> {
+impl<'a, 'b> StableXOrderBookReading for PaginatedStableXOrderBookReader<'a, 'b> {
     fn get_auction_index(&self) -> Result<U256> {
         self.contract
             .get_current_auction_index()
             .map(|batch_collecting_orders| batch_collecting_orders - 1)
     }
 
     fn get_auction_data(&self, index: U256) -> Result<(AccountState, Vec<Order>)> {
-        self.contract.get_auction_data(index)
+        let block = self.web3.eth().block_number().wait()?.as_u64();
+        let mut reader = BatchedAuctionDataReader::new(index);
+        loop {
```

Comment:
ubernit: Would a while loop be more elegant here?

Reply:
Don't think so. What I really want is a do-while loop but rust does not have them.

Reply:
The loop body would be empty since the condition would do the side-effect (we’d be checking apply_batch’s return value. Thus it shouldn’t matter if it’s while or do-while, shouldn’t it?

But maybe having the side effect in the condition is a bit ugly.